src/librustc/middle/mem_categorization.rs

```diff
@@ -1144,8 +1144,53 @@ impl<'a, 'gcx, 'tcx> MemCategorizationContext<'a, 'gcx, 'tcx> {
 
         debug!("cat_pattern: {:?} cmt={:?}", pat, cmt);
 
+        // FIXME(tschottdorf): I'm not clear on whether this should also see the adjusted `cmt`.
+        // I would assume so, but @nikomatsakis pointed me at just after this line to make the
+        // adjustments.
```

Comment:
I did point you to do the adjustment after, but I think I was wrong. I think we should move this callback to occur *after*. At first I thought we would do the callback both *before* and after each adjustment, but since we supply *both* the pattern *and* the cmt, I think it's better to just do a callback *after* the adjustments. We are therefore giving (a) the pattern that is being used and (b) the cmt that is being matched by this pattern. Looking at the use-sites of this function, all of them expect this and will work properly.

Reply:
Ok, will give that a try. Do you have an example or another way of testing that would break the current code and works with the callback invoked at the right place?

Reply:
mm let me think about it. There may be such a thing.

Reply:
I cannot find a way that it makes any difference -- all the consumers of this except for an obscure one in clippy only care about **binding patterns**, and those are never adjusted. However, it still seems more correct. Right now, we are passing in the bare pattern paired with the cmt that corresponds to (logically) the adjusted pattern. This is also *coherent* but it seems a bit .. less expected. i.e., if I match on the pattern and get (e.g.) `PatTuple`, I would expect the cmt to have tuple type (but in this code, it may have reference type).

Reply:
I just want to double check your comment above because the code that you commented on gives the `cmt` for the bare pattern and the bare pattern, whereas moving the callback gives the *adjusted* `cmt` and (still) the bare pattern. Your wording makes it sound like you'd actually prefer option 1, which is the current code (right?).

Reply:
So, what I meant is this:

- Consider first `match &Some(3) { &Some(ref x) => `.
    - When `cat_pattern` is first invoked, we have the cmt C for the discriminant (`&Some(3)`) and the pattern `&Some(ref x)`, and we invoke the callback with that
    - We then process the pattern, yielding the CMT for `*&Some(3)` and the pattern `Some(ref x)`, and we invoke the callback with *that*.
    - Finally, we process the CMT one step further to `Downcast<Some>(*&Some(3)).0` and invoke callback with with the pattern `ref x`.
- Under the code as it is there now:
    - When `cat_pattern` is first invoked, we still have the same cmt C for the discriminant (`&Some(3)`), but the pattern is for `Some(x)`. We invoke the callback with that.
    - Then we adjust the CMT to `*&Some(3)`. No callback occurs after this.
    - Then we process the pattern, downcasting to `Downcast<Some>(*&Some(3)).0`, and invoke callback with pattern `x`, where the ref is implied.

These don't quite match -- there aren't the same number of callbacks, to start, but also the cmt and patterns don't match up. If we move the callback to **after** the adjustment, we will get:

- Under the code as proposed:
    - When `cat_pattern` is first invoked, we still have the same cmt C for the discriminant (`&Some(3)`), but the pattern is for `Some(x)`.
    - Then we adjust the CMT to `*&Some(3)`.
    - Now we callback with CMT `*&Some(3)` and pattern `Some(x)`, just as before.
    - Then we process the pattern, downcasting to `Downcast<Some>(*&Some(3)).0`, and invoke callback with pattern `x`, where the ref is implied.